ArduCopter/mode.h

```diff
@@ -232,6 +232,67 @@ class Mode {
     };
     static AutoYaw auto_yaw;
 
+#if PRECISION_LANDING == ENABLED
+    // Precision Landing State Machine
+    class PrecLand_StateMachine {
+    public:
+        // Initialize the state machine. This is called everytime we switch to a mode that can do prec landing
+        void init();
+
+        // Run Prec Land State Machine. During Prec Landing, we might encouter four scenarios:
+        // 1. We had the target in sight, but have lost it now. 2. We never had the target in sight and user wants to land.
+        // 3. We have the target in sight and can continue landing. 4. The sensor is out of range
+        // This method deals with all of these scenarios
+        void update_precland_state_machine();
+
+    private:
+
+        // Target is lost (i.e we had it in sight some time back), this method helps decide on what needs to be done next
+        // The chosen action depends on user set landing strictness
+        void target_lost_actions();
+
+        // Retry landing based on a previously known location of the landing target
+        void retry_landing();
+
+        // Stop landing (hover) to a stopping location
+        void stop_landing();
+
+        // Decide what the next action is going to be if we have hit a failsafe
+        // Failsafe will only trigger as a last resort
+        void do_failsafe_actions();
+
+        // Reset the landing statemachines. This needs to be called everytime the landing target is back in sight.
+        // So that if the landing target goes out of sight again, we can start the failed landing procedure back from the beginning stage
+        void reset_failed_landing_statemachines();
+
+        // State machine for action to do when Landing target is lost (after it was in sight a while back)
+        enum class Landing_Target_Lost_Action: uint8_t {
+            INIT = 0,
+            DESCEND,
+            LAND_VERTICALLY,
+            RETRY_LANDING,
+        };
+        Landing_Target_Lost_Action landing_target_lost_action;  // Current action being done in the Lost Landing target state machine
+
+        // State Machine for landing retry
+        enum class RetryLanding : uint8_t {
+            RETRY_INIT = 0,
+            RETRY_IN_PROGRESS,
+            RETRY_COMPLETE
+        };
```

Comment:
```suggestion
        enum class RetryLandingState : uint8_t {
            INIT = 0,
            IN_PROGRESS,
            COMPLETE
        };
```

Would be nice to get a more complete description of these states - maybe just a sort `// foo` comment after each state